### PR TITLE
feat(probas,#13044): PyMC-16 exécute le GP sparse annoncé (HSGP, mur O(N³) mesuré, 2 budgets, N=1200)

### LINKED ISSUE #13044: PyMC-16 : exécuter réellement le GP sparse annoncé par le notebook
## Constat vérifié sur `origin/main`

Le notebook [`PyMC-16-Sparse-Gaussian-Process.ipynb`](https://github.com/jsboige/CoursIA/blob/main/MyIA.AI.Notebooks/Probas/PyMC/PyMC-16-Sparse-Gaussian-Process.ipynb) annonce un **processus gaussien sparse** dans son nom, ses renvois et sa prose, mais son modèle exécuté est un GP latent dense :

```python
cov = pm.gp.cov.ExpQuad(input_dim=2, ls=ls)
gp = pm.gp.Latent(cov_func=cov)
f = gp.prior("f", X=X_donut)
```

Le jeu d'entraînement contient seulement 16 points. Le notebook n'instancie ni points inducteurs, ni `pm.gp.MarginalSparse`/`MarginalApprox`, ni `pm.gp.HSGP`. Ces API apparaissent uniquement dans la section markdown « Pour aller plus loin ».

L'écart est pédagogique, pas seulement nominal :

- la lecture de la longueur de corrélation affirme que « le sparse du titre désigne l'astuce computationnelle (points induits) » ;
- la bibliographie et la synthèse développent le coût `O(NM²)` et les inducing points ;
- le registre twin `probas-16-sparse-gaussian-process.yaml` déclare un « socle pédagogique commun : gaussiens parcimonieux (Sparse GP, points induits) modelisé dans les deux stacks » ;
- le jumeau Infer.NET exécute effectivement `SparseGP` avec un basis set, puis compare 4 points inducteurs à un basis contenant tous les inputs ; le côté PyMC ne réalise aucune expérience équivalente.

Le notebook démontre correctement la **classification GP dense** avec NUTS, une frontière non linéaire et une longueur d'échelle apprise. Ce contenu reste utile. Le défaut est de présenter comme exécutée la capacité sparse qui demeure seulement citée.

## Pourquoi cela compte

Un étudiant ne peut pas vérifier dans l'état actuel :

- comment choisir le nombre ou l'emplacement des points inducteurs ;
- quel compromis précision/coût introduit l'approximation sparse ;
- si HSGP ou une approximation à points inducteurs reproduit la frontière du GP dense ;
- à partir de quelle taille le passage à l'échelle devient observable.

Le cas `N=16` masque précisément la propriété distinctive annoncée : le calcul dense est trivial et ne permet aucune comparaison de coût.

## Correction attendue

Conserver le bras dense existant comme baseline, puis ajouter un **vrai bras scalable PyMC** sur un problème assez grand pour rendre la différence observable :

1. choisir une API réellement disponible dans l'environnement du notebook (`pm.gp.HSGP` ou l'API sparse/approchée PyMC appropriée) ;
2. utiliser un dataset non linéaire plus grand, tout en gardant un temps d'exécution pédagogique borné ;
3. entraîner dense et sparse/scalable sur des splits comparables ;
4. mesurer au minimum le temps, la qualité prédictive hors échantillon et l'écart des probabilités/frontières ;
5. faire varier le budget d'approximation (nombre de bases/inducteurs ou rang HSGP) pour montrer le compromis ;
6. dériver la prose uniquement des outputs réels ;
7. synchroniser honnêtement le registre twin avec la capacité effectivement exécutée.

Si aucune API sparse n'est compatible avec la classification Bernoulli retenue, le notebook doit l'établir explicitement et être renommé/requalifié en classification GP dense ; la simple mention d'API de régression sparse ne suffit pas.

## Critères d'acceptation

- [ ] Un moteur GP sparse/scalable PyMC est **instancié et exécuté**, pas seulement cité en markdown.
- [ ] Le GP dense actuel est conservé comme baseline explicite.
- [ ] Le problème comporte assez d'observations pour exercer la différence de complexité sans rendre le notebook déraisonnablement lent.
- [ ] Au moins deux budgets d'approximation sont comparés.
- [ ] Temps d'exécution et métrique prédictive hors échantillon sont reportés pour les deux bras.
- [ ] Une visualisation compare les frontières ou probabilités prédictives dense vs sparse/scalable.
- [ ] La conclusion formule honnêtement le compromis observé et ses limites.
- [ ] Le registre twin ne déclare plus de capacité sparse commune non exécutée.
- [ ] Ré-exécution complète avec outputs réels, `execution_count` cohérents et zéro erreur volontaire (C.1/C.2).

## Déduplication

Les PR historiques sur ce notebook ont porté sur sa création, ses citations, la terminologie logit/probit, les sorties et l'interprétation de la longueur d'échelle. Aucun véhicule ouvert ne touche actuellement ce chemin et aucune issue retrouvée ne demande l'exécution d'un vrai bras sparse PyMC.

### PR DESCRIPTION
**Pull request by jsboige**
Grain: DEEP/notebook-python — lane myia-po-2026:CoursIA — prev: MED/guard #13143

Closes #13044

## Summary

PyMC-16 s'intitule « Sparse Gaussian Process » mais n'exécutait qu'un GP **dense** (`pm.gp.Latent`, N=16) — le « sparse » n'était cité qu'en conclusion. Cette PR ajoute une **section 6 exécutée** « Passer à l'échelle : le GP vraiment sparse (HSGP) » :

- **Mur O(N³) mesuré** : Cholesky de K_N pour N=200→6400 (0.6 ms → 1.4 s), exposant empirique ajusté (N^2.3 sur 4 points, théorie asymptotique N³).
- **Bras dense baseline N=200** conservé (`pm.gp.Latent` + Bernoulli logit, NUTS 200/200 ×2 chains, timed).
- **Bras HSGP ×2 budgets** N=200 : `pm.gp.HSGP(m=[10,10])` (100 coeffs) et `m=[20,20]` (400 coeffs), mêmes réglages NUTS, timed.
- **Bras échelle N=1200** : HSGP exécuté là où le dense est estimé hors budget (extrapolation Cholesky-dominée, hypothèse explicite).
- **OOS réel** : accuracy + log-loss sur 200 (puis 300) points de test pour chaque bras.
- **Figure 4 panneaux** (frontières postérieures + contour p=0.5) + **accord quantitatif** avec la référence dense (corrélation, |Δp| moyen).
- **Exercice 4** « budget HSGP trop petit » (m=[3,3]) ; synthèse et conclusion mises à jour.
- **README série** : règle E — fichier entier audité, 2 points dépendants corrigés (prose « inducing points … cellules ultérieures » devenue inexacte → HSGP exécuté section 6 ; durée 55 → 75 min, concepts).

## Métriques (sorties réelles des cellules, kernel python3 local)

| Bras | Temps | acc OOS | log-loss OOS |
|------|-------|---------|--------------|
| GP dense N=200 | 51.2 s | 1.000 | 0.113 |
| HSGP m=[10,10] N=200 | 3.3 s | 1.000 | 0.114 |
| HSGP m=[20,20] N=200 | 4.2 s | 1.000 | 0.113 |
| HSGP m=[20,20] N=1200 | 12.1 s | 1.000 | **0.030** |
| GP dense N=1200 (estimé) | ~51 min | — | — |

- Cholesky mesurée : 0.6 ms (N=200) → 12.5 (800) → 352 (3200) → 1443 ms (6400) ; ×32 points → ×2615 coût.
- Extrapolation Cholesky-dominée (hypothèse explicite dans le notebook) : dense N=1200 ≈ 51 min (rapport ~254× vs HSGP mesuré), N=6400 ≈ 40 h.
- Accord des frontières avec le dense : corr 0.999 (m=[10,10]) / 0.999 (m=[20,20]) / 0.989 (N=1200), |Δp| moyen 0.009 / 0.009 / 0.091.

## SOTA verdict

**SOTA-OK** — `pm.gp.HSGP` (PyMC 5.28.5, Solin & Särkkä 2020 / Ruitort-Mayol et al. 2023) réellement instancié et exécuté (sampling NUTS + `conditional` + `sample_posterior_predictive`). Le refus du dense à N=1200 est une **extrapolation d'une loi mesurée** (chaîne : temps dense N=200 mesuré × ratio Cholesky interpolé), pas un hand-wave.

Note technique (gotcha documenté dans le code) : le PPC grille d'un `conditional` Latent fait une SVD (G×G) par tirage — la grille comparative est à 30×30 (900 points) pour garder le coût borné ; la section originale 4 garde sa grille 20×20.

## Validation (H.1 / C.1 / C.2)

- Papermill in-place kernel `python3` : 38 cellules, 0 erreur, exec_counts 1..18 séquentiels (log d'exécution : 38/38 `Executing Cell`, exit 0).
- `validate_pr_notebooks.py` : **1/1 PASS** (18 code cells).
- `audit_c1_c3` : **1102/1102 pass** (aucun `raise NotImplementedError`/`assert False`/`1/0`).
- Scan machine-path sur outputs : **0 hit**.
- 4 figures générées réellement (dont le 4-panneaux comparatif).

## Twin registry

`probas-16-sparse-gaussian-process.yaml` : rebaseline via `check_twin_parity.py --update` en DERNIÈRE opération (python_sha 9302e5d1, csharp inchangé) + raison documentée + `known_differences` précisée : C# exécute le sparse par **points induits** (SparseGP basis set), Python par **base spectrale Hilbert** (HSGP) — deux familles d'approximation distinctes, toutes deux réellement exécutées. `--verify-recorded-sha` : 157/157 OK, 0 MISMATCH.

### CODE REVIEW

**Review by jsboige — COMMENTED** [review]
**[Hermes]** — review #13154 (DEEP/notebook PyMC, head `7754c3316c`, dedup : 0 review cluster sur ce SHA).

**Vérifications firsthand (diff 723 Ko trié par greps ciblés) :**

1. **Authenticité exécution** — `execution_count` 1→18 séquentiel, **0 null**. Les 4 figures sont de vraies images matplotlib (base64 `iVBOR` avec header versionné, metadata `matplotlib 3.10.x`).
2. **Claims du body vs outputs commités** — tous retrouvés dans les sorties : `51.2 s` dense ×2, `3.3 s` / `4.2 s` HSGP m=[10,10] / m=[20,20], `12.1 s` N=1200, `log-loss = 0.030` / `0.113` ×3 / `0.114`, échelle Cholesky `0.6 ms` → `12.5` → `352` → `1443 ms`, corr `0.999`/`0.999`/`0.989`, |Δp| `0.009`/`0.009`/`0.091`, extrapolations `51 min` (rapport `~254`) et `40 h`. Les m=[3,3] de l'exercice 4 sont bien présents (2 occurrences source).
3. **HSGP réellement instancié** — 5 occurrences `pm.gp.HSGP` sur lignes ajoutées (pas juste en prose), m=[10,10] ×14, m=[20,20] ×16. La section 6 exécute ce que le titre annonce — c'est bien le correctif de l'issue #13044.
4. **Security scan** — `HF_TOKEN|API_KEY|BEARER|PASSWORD|SECRET|TOKEN=` : **0 match**.
5. **0 stub** — `NotImplementedError|assert False` : 0 sur l'ensemble du diff (le claim 1102/1102 audit est cohérent).
6. **Twin registry** — rebaseline `probas-16-sparse-gaussian-process.yaml` : entrée datée 2026-08-26, `python_sha 9302e5d1`, `csharp_sha` inchangé — discipline twin respectée.
7. **README série** — durée 55→75 min et « probit »→« logit » corrigés en cohérence avec le bras dense Bernoulli-logit exécuté (la correction « probit → logit » est juste : le modèle du notebook est bien `pm.gp.Latent` + lien logit).

**Note (non-bloquant) :** pas de cellule `az.summary`/r_hat sur les nouveaux bras HSGP — les diagnostics de convergence NUTS ne sont pas affichés pour la section 6 (les bras denses de la section 4 les ont déjà). Si une re-exec ultérieure passe par là, un `az.summary(var_names=["__sqrt_PSD__", ...])` ou à minima un check de divergences renforcerait la lecture MCMC. Ne bloque pas : les acc/log-loss OOS accordés au dense (corr ≥0.989) sont le critère fonctionnel ici.

**Verdict : APPROVE sémantique (contrainte token : COMMENT only — auteur = jsboige).** Le mur O(N³) est mesuré, pas asserté ; le refus du dense à N=1200 est une extrapolation d'une loi ajustée, documentée comme hypothèse — honnête.
